src/chain/bls_emu.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Types emulating the BLS functionality until proper BLS lands
+use super::{delivery_group_size, NetworkEvent, ProofSet, SectionInfo};
+use crate::id::{FullId, PublicId};
+use parsec;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt,
+};
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct PublicKeySet {
+    sec_info: SectionInfo,
+    threshold: usize,
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct PublicKey(PublicKeySet);
+
+pub type SignatureShare = ::safe_crypto::Signature;
+
+pub struct SecretKeyShare(FullId);
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct PublicKeyShare(PublicId);
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct Signature {
+    sigs: BTreeMap<PublicId, SignatureShare>,
+}
+
+impl Signature {
+    pub fn from_proof_set(proofs: ProofSet) -> Self {
+        Self { sigs: proofs.sigs }
+    }
+}
+
+impl SecretKeyShare {
+    #[allow(unused)]
+    pub fn public_key_share(&self) -> PublicKeyShare {
+        PublicKeyShare(*self.0.public_id())
+    }
+
+    #[allow(unused)]
+    pub fn sign<M: AsRef<[u8]>>(&self, message: M) -> SignatureShare {
+        self.0.signing_private_key().sign_detached(message.as_ref())
+    }
+}
+
+impl PublicKeyShare {
+    #[allow(unused)]
+    pub fn verify<M: AsRef<[u8]>>(&self, sig: &SignatureShare, msg: M) -> bool {
+        self.0
+            .signing_public_key()
+            .verify_detached(sig, msg.as_ref())
+    }
+}
+
+impl PublicKeySet {
+    #[allow(unused)]
+    pub fn new(threshold: usize, keys: BTreeSet<PublicId>) -> Self {
+        let sec_info = SectionInfo::new(keys, Default::default(), None).unwrap();
+        Self::from_section_info(threshold, sec_info)
+    }
+
+    pub fn from_section_info(threshold: usize, sec_info: SectionInfo) -> Self {
+        Self {
+            threshold,
```

Comment:
In our use cases, it probably can - but I wanted to keep the code as generic as possible. This part probably won't make it to the code after BLS is merged, so maybe coupling the threshold with the section info here is not such a bad idea. What do you think @jeanphilippeD ?

Reply:
I guess you can make that function non public as well.
Since it is code that is going to go away, I'd go for whatever is simpler.

It seem it may be simplifying code & test if you use your delivery_group function inside it instead of taking the threshold param? would it?

Reply:
(as in you are using in in PublicKey, so why not moving it into PublicKeySet?)

Reply:
Yeah, I originally wanted to calculate the threshold completely outside this module to make the functionality boundaries more clear (BLS won't be aware of our threshold requirements, we'll have to pass it the threshold we want explicitly), but then I did this thing in `PublicKey`. I'm not sure which is better - make the method in `PublicKey` also take the threshold from outside, or just move this into `PublicKeySet`. Might be that it's the latter, as this code won't work with real BLS, anyway (we won't be able to derive keys from `SectionInfo`s, we'll have to get them from DKG).

Reply:
Correct me if I'm wrong, but we will not pass the threshold to either signature or public key, only the DKG.

The way we'll know we reach the threshold is if the combined signature is valid against the public key.

If anything, we would use the threshold directly in the signature validation. But where it is fine too (i.e pre-computed once and cached). and I don't want us spend too much time on something that is going to be disappearing soon-ish :-)

Reply:
(also I still assume we'll have to fix stuff when moving to real BLS).

Reply:
Yes, I think you're right - the threshold is passed to the DKG, then it is actually stored in the resulting `PublicKeySet` IIRC (it has to know what the degree of the polynomial needs to be when combining the signatures).